### DISCUSSION TITLE
y is not optional in ShapletTransform.fit(X, y)

### DISCUSSION THREAD

**Discussion by jacktang** [alan-turing-institute/sktime#6483] [Q&A] [answered]
Hello, I need ShapeletTransform to extract shapelets of X and read the doc [here]( https://www.sktime.net/en/stable/api_reference/auto_generated/sktime.transformations.panel.shapelet_transform.ShapeletTransform.html#sktime.transformations.panel.shapelet_transform.ShapeletTransform.fit), it said the `y` is optional, and it is not. I read the source code, here is the reference of `y`:

https://github.com/sktime/sktime/blob/a9848dcdbed47721293f5e5e743cb291f5917ff7/sktime/transformations/panel/shapelet_transform/_shapelet_transform.py#L181

Should it change to 
```
num_ins = X.shape[0]
```
?

My test code lists below:

```
from sktime.transformations.panel.shapelet_transform._shapelet_transform import ShapeletTransform
import pandas as pd

X = pd.DataFrame([[0, 2, 3, 4, 3, 2, 1],
      [0, 1, 3, 4, 3, 4, 5],
      [2, 1, 0, 2, 1, 5, 4],
      [1, 2, 2, 1, 0, 3, 5]])
y = [0, 0, 1, 1]

st = ShapeletTransform()
st._fit(X, y) # what I really want is `st._fit(X)`
```

**Comment by fkiraly** [comment]
@jacktang, unfortunately `y` is not optional in either of the shapelet transforms, at the moment.

The docs are wrong, in this case.

The requirement is because the algorithms (`ShapeletTransform` and `ShapeletTransformPyts`) compute not just shapelet coefficients, but select shapelets that are maximally discriminative between the classes. If there are on classes, "discriminative" makes no mathematical sense.

If you want shapelet-like features in general, you could try [the rocket family of transformations](https://www.sktime.net/en/latest/api_reference/transformations.html#shapelets-wavelets-and-convolution) which is basically a random shapelet/kernel approach and does not require `y`.

**Reply by fkiraly** [reply]
Tried to improve docstrings here, this also makes clear how to know if `y` is required: https://github.com/sktime/sktime/pull/6489

**Reply by jacktang** [reply]
Thanks for the reply! As you see the example, I provided `y` values and the code was still failed.

**Reply by fkiraly** [reply, marked as answer]
oh, I see. The code is not valid, you need to provide `X` and `y` in the formats specified in the docstring, and not call `_fit` directly but `fit`.

Example code that works, assuming you want `X` to represent 4 time series with 1 variable and 7 time points each:

```python
from sktime.transformations.panel.shapelet_transform._shapelet_transform import ShapeletTransform
import numpy as np
import pandas as pd

X = np.array([[0, 2, 3, 4, 3, 2, 1],
      [0, 1, 3, 4, 3, 4, 5],
      [2, 1, 0, 2, 1, 5, 4],
      [1, 2, 2, 1, 0, 3, 5]]).reshape((4, 1, 7))
y = np.array([0, 0, 1, 1])

st = ShapeletTransform()
st.fit(X, y)
```

**Reply by jacktang** [reply]
Thank you! it helps a lot :)
